### PR TITLE
feat: add DSearch API

### PR DESCRIPTION
**Pull request by Lanttcat**
## Description

Closes #mf-2822
## Type of change

- [ ] Documentation
- [ ] Code refactoring (Restructuring existing code w/o changing its observable behavior)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (a fix or feature that would make something no longer possible to do/require old user must upgrade their Mask Network to this new version)

## Checklist

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
  - [ ] I have removed all in development `console.log`s
  - [ ] I have removed all commented code.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have read [Internationalization Guide](https://github.com/DimensionDev/Maskbook/blob/develop/docs/i18n-guide.md) and moved text fields to the i18n JSON file.

If this PR depends on external APIs:

- [ ] I have configured those APIs with CORS headers to let extension requests get passed. 
  - chrome extension: `chrome-extension://[id]`
  - firefox extension: `moz-extension://[id]`
- [ ] I have delegated all web requests to the background service via the internal RPC bridge.

### CODE REVIEW

File: packages/web3-providers/src/DSearch/index.ts [hunk 1]
```diff
@@ -1,19 +1,207 @@
...
+                    type: 'fuzzy',
+                    fullSearch: (keyword: string, all: Resource[]) => {
+                        const data = all.map((x) => ({
+                            ...x,
+                            sSymbol: x.symbol?.replace(/\s/g, ''),
+                            sName: x.name?.replace(/\s/g, ''),
+                        }))
+                        console.log(data)
+
+                        const fuse = new Fuse(data, {
```
Comment (guanbinrui) [inline packages/web3-providers/src/DSearch/index.ts]: There is FuseAPI in `@masknet/web3-provider`.

File: packages/web3-providers/tests/DSearch/index.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
+import { describe, expect, test } from 'vitest'
+import { DSearchAPI } from '../../src/DSearch/index.js'
+
+describe('DSearch test', () => {
+    test('should return by sysmbol from specific list', async () => {
+        const DSearch = new DSearchAPI()
+        const result = await DSearch.search('token:eth')
```
Comment (guanbinrui) [inline packages/web3-providers/tests/DSearch/index.ts:9]: Does the perfix `token:` from the user?

File: packages/web3-providers/src/DSearch/index.ts [hunk 3]
```diff
@@ -1,19 +1,207 @@
...
+        telegram?: string
+        github?: string
+        instagram?: string
+        medium?: string
+    }
+}
+
+export type Resource = FungibleToken | NonFungibleCollectionRemote
+
+type rule = {
```
Comment (guanbinrui) [inline packages/web3-providers/src/DSearch/index.ts]:
```suggestion
interface Rule {
```

File: packages/web3-providers/src/DSearch/index.ts [hunk 4]
```diff
@@ -1,19 +1,207 @@
...
+    key: string
+    type: 'exact' | 'fuzzy'
+    filter?: (data: Resource, keyword: string, all: Resource[]) => boolean
+    fullSearch?: (keyword: string, all: Resource[]) => Resource[]
+}
+type handler = {
```
Comment (guanbinrui) [inline packages/web3-providers/src/DSearch/index.ts]:
```suggestion
interface Handler {
```

File: packages/web3-providers/src/DSearch/index.ts [hunk 5]
```diff
@@ -1,19 +1,135 @@
...
+    NFTScanClient: DSearchBaseAPI.DataSourceProvider<ChainId, SchemaType>
+    CoinGeckoClient: DSearchBaseAPI.DataSourceProvider<ChainId, SchemaType>
+    CoinMarketCapClient: DSearchBaseAPI.DataSourceProvider<ChainId, SchemaType>
+
+    constructor() {
+        this.handlers = getHandlers<ChainId, SchemaType>()
+        this.NFTScanClient = new NFTScanSearchAPI<ChainId, SchemaType>()
+        this.CoinGeckoClient = new CoinGeckoSearchAPI<ChainId, SchemaType>()
+        this.CoinMarketCapClient = new CoinMarketCapSearchAPI<ChainId, SchemaType>()
+    }
```
Comment (guanbinrui) [inline packages/web3-providers/src/DSearch/index.ts]:
```suggestion
    handlers = getHandlers<ChainId, SchemaType>()

    NFTScanClient = new NFTScanSearchAPI<ChainId, SchemaType>()
    CoinGeckoClient = new CoinGeckoSearchAPI<ChainId, SchemaType>()
    CoinMarketCapClient = new CoinMarketCapSearchAPI<ChainId, SchemaType>()
```

File: packages/web3-shared/base/src/specs/index.ts [hunk 6]
```diff
@@ -609,18 +607,42 @@ export interface DomainResult<ChainId> extends Result<ChainId> {
...
+    source: SourceType 
     token?: FungibleToken<ChainId, SchemaType>
 }
 
 export interface NonFungibleTokenResult<ChainId, SchemaType> extends Result<ChainId> {
+    name: string,
     address: string
-    tokenId: string
+    tokenId?: string
+    chain?: ChainId
```
Comment (guanbinrui) [inline packages/web3-shared/base/src/specs/index.ts]: Dup with `chainId`.

File: packages/web3-shared/base/src/specs/index.ts [hunk 7]
```diff
@@ -593,10 +593,8 @@ export interface FungibleTokenSecurity {}
 export interface NonFungibleTokenSecurity {}
 
  export interface Result<ChainId> {
-    pluginID: NetworkPluginID
-    chainId: ChainId
+    chainId?: ChainId
```
Comment (guanbinrui) [inline packages/web3-shared/base/src/specs/index.ts:607]: The `pluginID` is required to distinguish `chainId`.

File: packages/web3-shared/base/src/specs/index.ts [hunk 8]
```diff
@@ -609,18 +607,42 @@ export interface DomainResult<ChainId> extends Result<ChainId> {
...
+    type: SearchResultType.NonFungibleToken
+    source: SourceType 
     token?: NonFungibleToken<ChainId, SchemaType>
 }
 
 export interface NonFungibleCollectionResult<ChainId, SchemaType> extends Result<ChainId> {
+    name: string,
     address: string
+    tokenId?: string
+    chain?: ChainId
```
Comment (guanbinrui) [inline packages/web3-shared/base/src/specs/index.ts]: Duplicate field.

File: packages/web3-shared/base/src/specs/index.ts [hunk 9]
```diff
@@ -609,18 +607,42 @@ export interface DomainResult<ChainId> extends Result<ChainId> {
...
+    source: SourceType 
+    type: SearchResultType.NonFungibleCollection
+    socialLinks?: {
```
Comment (guanbinrui) [inline packages/web3-shared/base/src/specs/index.ts]: Add it to `NonFungibleCollection`.

File: packages/web3-shared/base/src/specs/index.ts [hunk 10]
```diff
@@ -609,18 +607,42 @@ export interface DomainResult<ChainId> extends Result<ChainId> {
...
+    chain?: ChainId
+    type: SearchResultType.NonFungibleToken
+    source: SourceType 
     token?: NonFungibleToken<ChainId, SchemaType>
 }
 
 export interface NonFungibleCollectionResult<ChainId, SchemaType> extends Result<ChainId> {
+    name: string,
     address: string
+    tokenId?: string
```
Comment (guanbinrui) [inline packages/web3-shared/base/src/specs/index.ts]: Why a collection has got `tokenId` field.

File: packages/web3-shared/base/src/specs/index.ts [hunk 11]
```diff
@@ -609,18 +607,42 @@ export interface DomainResult<ChainId> extends Result<ChainId> {
 }
 
 export interface FungibleTokenResult<ChainId, SchemaType> extends Result<ChainId> {
-    address: string
+    //  This id on the provider platform
+    id?: string
```
Comment (guanbinrui) [inline packages/web3-shared/base/src/specs/index.ts:623]: Why UI needs this?

Comment (Lanttcat) [reply packages/web3-shared/base/src/specs/index.ts:623]: The DSearch API return the tidy data. The caller need to get detail from data provider by id.

**Review by guanbinrui — APPROVED** [review]

**Pull request merged**